Add a stub implementation of MediaPlayerPrivateWirelessPlayback

aestes opened the pull request:
#### afdb9152dce617ca99b8e6f4b344ae53c832cf40
<pre>
Add a stub implementation of MediaPlayerPrivateWirelessPlayback
<a href="https://bugs.webkit.org/show_bug.cgi?id=302140">https://bugs.webkit.org/show_bug.cgi?id=302140</a>
<a href="https://rdar.apple.com/164226183">rdar://164226183</a>

Reviewed by Jean-Yves Avenard.

Added MediaPlayerPrivateWirelessPlayback with stub implementations of pure virtual functions in
MediaPlayerPrivateInterface. Added MediaPlayerFactoryWirelessPlayback for creating instances of
MediaPlayerPrivateWirelessPlayback and registered it as a media engine in MediaPlayer when the
WirelessPlaybackMediaPlayerEnabled feature is enabled.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/DeprecatedGlobalSettings.h:
(WebCore::DeprecatedGlobalSettings::setWirelessPlaybackMediaPlayerEnabled):
(WebCore::DeprecatedGlobalSettings::isWirelessPlaybackMediaPlayerEnabled):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::WTF_REQUIRES_LOCK):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivateWirelessPlayback.cpp: Added.
(WebCore::MediaPlayerPrivateWirelessPlayback::registerMediaEngine):
(WebCore::MediaPlayerPrivateWirelessPlayback::MediaPlayerPrivateWirelessPlayback):
(WebCore::MediaPlayerPrivateWirelessPlayback::logChannel const):
* Source/WebCore/platform/graphics/MediaPlayerPrivateWirelessPlayback.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::supportsLinearMediaPlayer const):

Canonical link: <a href="https://commits.webkit.org/302774@main">https://commits.webkit.org/302774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2d0c1b18ff33514851ab93ca559ae204433eb47

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130035 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2296 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40896 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137430 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81549 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e423a746-49a7-4199-ae2b-3b67ca99a35c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131906 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2261 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2188 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99054 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66857 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/96aacd08-8866-49ed-87e1-93797a49aa95) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132982 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1681 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116468 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79755 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3f4ac86d-737e-494c-a057-776d6b2205c6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34593 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80702 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122027 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110128 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35103 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139909 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128487 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2090 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1938 "Found 1 new test failure: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/selectors/cue/white-space_normal_wrapped.html (failure)") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2135 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112815 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27376 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1669 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31266 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54923 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2163 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65532 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161501 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1979 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40257 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2012 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2086 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/MediaPlayerPrivateWirelessPlayback.h:
```diff
@@ -0,0 +1,106 @@
+#if !RELEASE_LOG_DISABLED
+    , private LoggerHelper
+#endif
+    , public CanMakeWeakPtr<MediaPlayerPrivateWirelessPlayback>
```
use RefCountedAndCanMakeWeakPtr ?

jyavenard commented on Source/WebCore/page/DeprecatedGlobalSettings.h:
```diff
@@ -126,6 +126,11 @@ class DeprecatedGlobalSettings {
 #endif
 
+#if ENABLE(WIRELESS_PLAYBACK_MEDIA_PLAYER)
+    static void setWirelessPlaybackMediaPlayerEnabled(bool isEnabled) { singleton().m_wirelessPlaybackMediaPlayerEnabled = isEnabled; }
```
We probably can do without those, and use plain settings instead with a combination of MediaStrategy/LoadOptions

aestes replied:
I will look into that in a follow-up.

jyavenard commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -362,6 +366,15 @@ static void buildMediaEnginesVector() WTF_REQUIRES_LOCK(mediaEngineVectorLock)
 #endif
 
+#if ENABLE(WIRELESS_PLAYBACK_MEDIA_PLAYER)
+    if (DeprecatedGlobalSettings::isWirelessPlaybackMediaPlayerEnabled()) {
```
We don't want any MediaPlayer in the GPU process going forward.

Can we build it to work from the content process only instead ?

aestes replied:
Possibly, but I haven't gotten that far in the design. I'll leave this as-is for now and simplify it later if possible.

aestes committed:
Add a stub implementation of MediaPlayerPrivateWirelessPlayback
https://bugs.webkit.org/show_bug.cgi?id=302140
rdar://164226183

Reviewed by Jean-Yves Avenard.

Added MediaPlayerPrivateWirelessPlayback with stub implementations of pure virtual functions in
MediaPlayerPrivateInterface. Added MediaPlayerFactoryWirelessPlayback for creating instances of
MediaPlayerPrivateWirelessPlayback and registered it as a media engine in MediaPlayer when the
WirelessPlaybackMediaPlayerEnabled feature is enabled.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/DeprecatedGlobalSettings.h:
(WebCore::DeprecatedGlobalSettings::setWirelessPlaybackMediaPlayerEnabled):
(WebCore::DeprecatedGlobalSettings::isWirelessPlaybackMediaPlayerEnabled):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::WTF_REQUIRES_LOCK):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivateWirelessPlayback.cpp: Added.
(WebCore::MediaPlayerPrivateWirelessPlayback::registerMediaEngine):
(WebCore::MediaPlayerPrivateWirelessPlayback::MediaPlayerPrivateWirelessPlayback):
(WebCore::MediaPlayerPrivateWirelessPlayback::logChannel const):
* Source/WebCore/platform/graphics/MediaPlayerPrivateWirelessPlayback.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::supportsLinearMediaPlayer const):

Canonical link: https://commits.webkit.org/302774@main
